geode/dispatcher.py

```diff
@@ -149,34 +150,37 @@ def __init__(self, config=None):
     async def distance_rows(self, missing, sem, session=None, provider=None):
         odim = [0, 1]
         ddim = [2, 3]
-        if missing.groupby(level=odim).ngroups < missing.groupby(level=ddim).ngroups:
+        if missing.groupby(level=odim).ngroups <= missing.groupby(level=ddim).ngroups:
+            idx = missing.groupby(level=odim, sort=False)
             res = await asyncio.gather(*[
                 self.throttled_distance_matrix(
                     origins=[o],
                     destinations=ds.index.to_frame(index=False).values[:, ddim],
                     sem=sem,
                     session=session,
                     provider=provider
-                ) for o, ds in missing.groupby(level=odim)
+                ) for o, ds in idx
             ])
         else:
+            idx = missing.groupby(level=ddim, sort=False)
             res = await asyncio.gather(*[
                 self.throttled_distance_matrix(
                     origins=ds.index.to_frame(index=False).values[:, odim],
                     destinations=[o],
                     sem=sem,
                     session=session,
                     provider=provider
-                ) for o, ds in missing.groupby(level=ddim)
+                ) for o, ds in idx
             ])
 
-
         res_df = None
         res_flat = [row.distances.ravel() for row in res if len(row.distances)]
         if res_flat:
             res_df = pd.DataFrame.from_records(
                 np.concatenate(res_flat),
-                index=missing.index)
+                # make new index by flat iterate through the groupby index
+                index=pd.MultiIndex.from_tuples((d for _, ds in idx for d in ds.index), names=missing.index.names)
```

Comment:
important fix of the PR, destination-major traversal with > 1 destination would fail.
as would any index supplied not in origin-major order (possible situation when called from distance_pairs endpoint).